Adds "failOnSterr: true" to CI steps and fixes bug in TakensEmbedding

ulupo opened the pull request:
#### Reference Issues/PRs
(#3)


#### What does this implement/fix? Explain your changes.
Currently, even when `pytest` tests and `flake8` tests fail, the azure pipelines will display a tick and continue. This PR began as an attempt to find a solution to this by adding the option `failOnStderr: false` to as many steps in the Azure CI as possible. Meanwhile, the `azure-pipelines.yml` file and `.azure-ci/docker_scripts.sh` were reorganized for clarity.

However, in the process, the following additional issues were noticed requiring patches and the opportunity was taken to fix them:

1. Executing the jupyter notebooks in `examples` with `jupyter nbconvert` was not working as desired. The new solution uses `papermilll` and prints cell outputs to the console. It appears to correctly print to stderr and to fail on Azure when exceptions are raised in executing cells. However *note*: warnings are also printed to stderr and hence will make this step fail when they occur.
2. As a byproduct of 1, it was noticed that many runtime warnings are not typically printed in jupyter notebooks when `joblib` parallelism is employed. On the other hand, `papermill` does catch these. Tracking down one such warning involving `TakensEmbedding` revealed a bug with the implementation of the false nearest neighbor criterion to choose optimal embedding dimensions. This bug was fixed in 4dc3e81aae3fae097500b4a78c025be41e2aca1d. Another issue was the use of `stride=1` regardless of the user-set `self.stride` in the call to the false nearest neighbour algorithm in `fit`. This was fixed in 682ed43.
3. A mistake in the `twine upload` steps in `azure-pipelines.yml` where `giotto-tda` was passed instead of the correct user `giott-learn` was fixed.
4. Code coverage generation from testing installed wheels was dropped from the build-type Azure jobs, as it was not working anyway but coverage is already generated when running `pytest` on the dev installation in previous steps.
5. Code cleanup and linting was performed in various places, including some config files (`doc/conf.py`, `setup.cfg`) and the jupyter notebooks in `examples`.


#### Any other comments?
1. There are a few places where `failOnStderr` is still set to `false` as we do not know how yet how to silence seemingly (?) unimportant warnings which would be printed to stderr.
2. It is not clear that the original objective has been fully reached. There still seems to be some randomness or uncertainty as to whether failed tests really show up as crosses in Azure.
3. It was noticed that *fewer* tests are ran when the installed wheels are tested vs when the dev installation is tested. It should be fully understood whether this is the expected behaviour.

Commit:
Add failOnSterr: true to testing and linting steps

Also change false to true in existing instance of failOnSterr

Commit:
Increase timeout threshold for nbconvert

classifying_shapes seems to require this

Commit:
Add line to upgrade nbconvert

Commit:
Add --ExecutePreprocesspr.enabled=True in nbconvert line

Commit:
Upgrade jupyter in azure pipelines

Commit:
Add pandas install to testing part of azure pipelines

It is required by mapper_quickstart.ipynb

Commit:
Try with a configuration file

Testing hypothesis that nbconvert is failing because of concurrent execution between notebooks

Commit:
Typo fix and note on pandas in mapper_quickstart

Commit:
Try to fix use of config file for nbconvert

Commit:
Fix azure-pipelines in view of previous config change

Commit:
More precise configuration details

Commit:
Fix incorrect attribute

Commit:
Place equal sign after --config

Commit:
Revert to command line, use startup_timeout

Commit:
Set infinite timeout limits

Commit:
Set timeouts to 1h

Commit:
Place lorenz attractor notebook in subfolder

Commit:
Hide another two notebooks

Commit:
Hide all notebooks and create small test one

Commit:
Try with papermill instead of nbconvert

Commit:
Change dash for underscore, remove execution-timeout

Commit:
Use backslash in windows papermill line

Commit:
Try with two backslashes

Commit:
Specify azure bundle in pip install papermill

Commit:
Navigate in and out of examples in windows

Commit:
Test with bash

Commit:
Remove extra space

Commit:
Test that pipelines *fail* correctly

Commit:
Remove nbconvert conf file

Commit:
Put original notebooks back, remove test notebook

Commit:
Try with a for loop in bash in windows

Commit:
Correctly use for variable in bash

Commit:
Try moving inside examples before running papermill

ulupo committed:
Merge pull request #10 from ulupo/master

Merge master into azure-ci/make_failed_tests_fail_pipelines

Commit:
Experiment with --log-level papermill option

Python warnings are printed to stderr currently, this is an attempt fix.

Commit:
Try papermill with script instead of bash

Commit:
Adapt to Windows command line

ulupo committed:
Attempt to fix papermill script

Change %notebook to %n in for loop

ulupo committed:
Try with %%

Commit:
Hide Pipeline output from Lorenz notebook

Commit:
More polishing of Lorenz notebook

Commit:
More polishing of notebooks

Commit:
Grammar fix in classifying_shapes.ipynb

Commit:
Avoid use of "is" with string literals

This should resolve some SyntaxWarnings when flake8 is run

Commit:
Directly import classes in lorenz_attractor.ipynb

Commit:
Merge remote-tracking branch 'origin/azure-ci/make_failed_tests_fail_pipelines' into azure-ci/make_failed_tests_fail_pipelines

Commit:
Refactor all test pipelines

Commit:
Reinstate brew install steps for macOS, remove some parentheses

Commit:
Swapped macOS and ubuntu tests by accident

Commit:
Minor changes in docker_scripts.sh

- Replace ".[tests, doc]" with ."[tests, doc]"
- Remove twine install step as it is not used here

Commit:
Replace giotto-tda with giotto-tda-nightly in PyPI upload step

This should reflect the fact that this step is only executed when 'nightly_check' is false

Commit:
Remove seemingly unnecessary 'Install pytest suite' step from linux

All operations here are presumably already executed by the bash script called by the previous step.

Commit:
Replace giotto-tda with giotto-tda-nightly in macOS PyPI upload

Also mode pip install twine step to here as it is not needed earlier

Commit:
Refactor macOS build pipeline

Commit:
Change giotto-tda to giotto-tda-nightly in wheel upload step for win

As the 'nightly_check' variable is supposed to be true, this seems appropriate

Commit:
Refactor windows build recipe

Commit:
Merge branch 'master' into azure-ci/make_failed_tests_fail_pipelines

Commit:
Remove spurious failOnStderr

Commit:
Add cell for ignoring runtime warnings in lorenz_attractor notebook

Commit:
Fix typos

Commit:
Fix some errors in classifying_shapes.ipynb

Commit:
Try a logging approach to avoid printing warnings to stderr

Commit:
Add brew install cmake for consistency with README instructions

Commit:
Remove remaining "hl." from Lorenz notebook

Commit:
Add brew update cmake step

Commit:
Hide lorenz attractor notebook as test

Commit:
brew update -> brew upgrade

Commit:
Reinsert Lorenz notebook

Commit:
Delete cell ignoring warnings from Lorenz notebook

Commit:
Clean TakensEmbedding._false_nearest_neighbors

Improve code readability and effectively silence division by zero warnings by setting results of division by zero to zero

Commit:
Replace giotto-tda-nightly with giotto-learn

-u denotes the user name, and since the PyPI user is giotto-learn, this is what has to be in this line. Thanks @matteocao for pointing this out!

Commit:
Remove wheel building step from docker_scripts

Commit:
Remove uninstall dev step from docker_scripts

Commit:
Refactor build pipelines

In manylinux case, try to move as much as possible away from docker_scripts. Also, attempt fix at jupyter notebooks test scripts.

Commit:
Set failOnStderr to false in wheel building step

Commit:
Set failOnStderr to false in Linux docker, use bash in win wheel install

Commit:
Place pip uninstall step back in docker_scripts

Commit:
Add dot to pytest command line argument

Commit:
Reinsert testin, linting and wheel building step in docker_scripts

Commit:
Forgot to remove wheel install step for manylinux from azure-pipelines

Commit:
Return manylinux CI to a previous state

Commit:
Try another path for notebook testing

Commit:
Fix cd'ing in jupyter notebook tests

Commit:
Allow failures of pytest on installed wheels

Commit:
Pass self.stride instead of always 1 to _false_nearest_neighbors!

Commit:
Adopt uniform standard for dev installation

Commit:
Add failOnStderr to test_win2016 dev install step

Commit:
Set failOnStderr to true in renaming steps

Commit:
Add failOnStderr to sphinx build step

wreise commented on azure-pipelines.yml:
```diff
@@ -337,51 +405,69 @@ jobs:
       sed -i $'s/\r$//' README.rst
+      pip install wheel
       python setup.py sdist bdist_wheel
+    failOnStderr: false
```
Inspect azure pipelines, to see if/why `False` is needed.

Commit:
Create local temporary directory to avoid permission issues

Also retrigger failOnStderr in wheel/coverage steps and wheel install steps

Commit:
Remove unnecessary comments on check_is_fitted

Commit:
Fix TakensEmbedding unit test which failed following 682ed43fbf61fee20fac47cd54d092802fd86b36

Commit:
Turn off failOnStderr for macOS

Commit:
Change pytest command in windows, comment out rest

Commit:
Remove final coverage report generation from build jobs

This is redundant as it is already done on the dev install in previous steps and in any case we cannot yet make it work. The wheel testing remains an important step.

Commit:
Remove deprecated autodoc_default_flags

Substitute with autodoc_default_options

Commit:
Change autodoc_default_option value to dict

Commit:
Merge branch 'master' into azure-ci/make_failed_tests_fail_pipelines

Commit:
Add a pytest.ini file to silence junit_family warning

Commit:
Pass --no-cov=True in wheel testing steps when building

Commit:
Delete pytest.ini and place junit_family in setup.cfg

Commit:
Avoid brew error when package is already up-to-date

Commit:
Merge branch 'master' into azure-ci/make_failed_tests_fail_pipelines

Commit:
Remove "=True" from --no-cov

Commit:
No longer try to upgrade cmake with brew

Commit:
Merge branch 'master' into azure-ci/make_failed_tests_fail_pipelines

Commit:
Update step description

Commit:
Merge branch 'master' into azure-ci/make_failed_tests_fail_pipelines

Commit:
Add --no-coverage-upload option to pytest for wheels

Commit:
Merge branch 'master' into azure-ci/make_failed_tests_fail_pipelines
